### PR TITLE
expresion, executor: correct the type inference of function "sum" and "avg"

### LINKED ISSUE #5494: interface conversion: interface {} is nil, not *types.MyDecimal
## 1. What did you do?
```sql
drop table if exists t;
create table t(id int primary key, b varchar(50), c int);
insert into t values(1, '1ff', NULL), (2, '234.02', 1);
select sum(b) from t;
```

## 2. What did you expect to see?

```sql
MySQL(localhost) > select sum(b) from t;
+--------+
| sum(b) |
+--------+
| 235.02 |
+--------+
1 row in set (0.00 sec)
```

## 3. What did you see instead?

```sql
TiDB(localhost) > select sum(b) from t;
ERROR 2006 (HY000): MySQL server has gone away
No connection. Trying to reconnect...
Connection id:    3
Current database: test

ERROR 2013 (HY000): Lost connection to MySQL server during query
```

log of tidb-server: 

```txt
2017/12/26 11:07:45.759 conn.go:410: [error] lastCmd select sum(b) from t, interface conversion: interface {} is nil, not *types.MyDecimal, goroutine 483 [running]:
github.com/pingcap/tidb/server.(*clientConn).Run.func1(0xc421441ad0, 0xc421471e0f)
        /Users/jianzhang.zj/opt/gopath/src/github.com/pingcap/tidb/server/conn.go:408 +0x10e
panic(0x1cd2de0, 0xc4214ca640)
        /Users/jianzhang.zj/opt/go-1.9/src/runtime/panic.go:491 +0x283
github.com/pingcap/tidb/types.(*Datum).GetMysqlDecimal(...)
        /Users/jianzhang.zj/opt/gopath/src/github.com/pingcap/tidb/types/datum.go:235
github.com/pingcap/tidb/types.DatumRow.GetMyDecimal(...)
        /Users/jianzhang.zj/opt/gopath/src/github.com/pingcap/tidb/types/row.go:138
github.com/pingcap/tidb/types.(*DatumRow).GetMyDecimal(0xc4214b1a40, 0x0, 0xc4214bb800)
        <autogenerated>:1 +0x88
github.com/pingcap/tidb/server.dumpTextRow(0xc4214bbc00, 0x4, 0x400, 0xc42054afc8, 0x1, 0x1, 0x25fec40, 0xc4214b1a40, 0xc420e0ff00, 0x1, ...)
        /Users/jianzhang.zj/opt/gopath/src/github.com/pingcap/tidb/server/util.go:306 +0xc5b
github.com/pingcap/tidb/server.(*clientConn).writeResultset(0xc421441ad0, 0x2e41588, 0xc42148bbc0, 0x25e4a80, 0xc4214d0240, 0xc420e00000, 0x0, 0x0)
        /Users/jianzhang.zj/opt/gopath/src/github.com/pingcap/tidb/server/conn.go:893 +0x3c9
github.com/pingcap/tidb/server.(*clientConn).handleQuery(0xc421441ad0, 0x2e41588, 0xc42148bbc0, 0xc4201a2e41, 0x14, 0x0, 0x0)
        /Users/jianzhang.zj/opt/gopath/src/github.com/pingcap/tidb/server/conn.go:810 +0x115
github.com/pingcap/tidb/server.(*clientConn).dispatch(0xc421441ad0, 0xc4201a2e41, 0x15, 0x15, 0x0, 0x0)
        /Users/jianzhang.zj/opt/gopath/src/github.com/pingcap/tidb/server/conn.go:592 +0x595
github.com/pingcap/tidb/server.(*clientConn).Run(0xc421441ad0)
        /Users/jianzhang.zj/opt/gopath/src/github.com/pingcap/tidb/server/conn.go:452 +0x1b8
github.com/pingcap/tidb/server.(*Server).onConn(0xc420840b40, 0x25e7600, 0xc420e512f0)
        /Users/jianzhang.zj/opt/gopath/src/github.com/pingcap/tidb/server/server.go:316 +0x144
created by github.com/pingcap/tidb/server.(*Server).Run
        /Users/jianzhang.zj/opt/gopath/src/github.com/pingcap/tidb/server/server.go:262 +0x496

2017/12/26 11:07:45.759 server.go:298: [info] [3] close connection
```

## 4. What version of TiDB are you using (`tidb-server -V` or run `select tidb_version();` on TiDB)?

```txt
$./bin/tidb-server -V
Release Version: v1.1.0-alpha-338-g100bc58a8
Git Commit Hash: 100bc58a833bcf738537fe1e559c325ae73c3b5d
Git Commit Branch: master
UTC Build Time:  2017-12-26 03:05:50
```

### PR DESCRIPTION
**Pull request by zz-jason**
fix #5494

### CODE REVIEW

**Comment by zz-jason** [comment]
@XuHuaiyu @hanfei1991 PTAL

File: expression/aggregation/avg.go [hunk 1]
```diff
@@ -37,10 +37,8 @@ func (af *avgFunction) Clone() Aggregation {
 // GetType implements Aggregation interface.
 func (af *avgFunction) GetType() *types.FieldType {
 	var ft *types.FieldType
-	if types.IsTypeFloat(af.Args[0].GetType().Tp) {
-		ft = types.NewFieldType(mysql.TypeDouble)
-		ft.Decimal = af.Args[0].GetType().Decimal
-	} else {
+	switch af.Args[0].GetType().Tp {
```
Comment (winoros) [inline expression/aggregation/avg.go]: add comments for this switch branch.

File: expression/aggregation/sum.go [hunk 2]
```diff
@@ -68,12 +68,13 @@ func (sf *sumFunction) CalculateDefaultValue(schema *expression.Schema, ctx cont
 // GetType implements Aggregation interface.
 func (sf *sumFunction) GetType() *types.FieldType {
 	var ft *types.FieldType
-	if types.IsTypeFloat(sf.Args[0].GetType().Tp) {
-		ft = types.NewFieldType(mysql.TypeDouble)
-	} else {
+	switch sf.Args[0].GetType().Tp {
```
Comment (winoros) [inline expression/aggregation/sum.go]: ditto

**Comment by coocood** [comment]
LGTM

**Pull request merged**